applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,325 @@
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_POSITION UCHAR_MAX
+#define DEBOUNCE_DELAY K_MSEC(5)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+	STATE_ERROR,
+};
+
+struct selector {
+	const struct selector_config *config;
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t position;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(selector_config)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	__ASSERT_NO_MSG((selector->position < selector->config->pins_size) ||
+			(selector->position == DEFAULT_POSITION));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_id = selector->config->id;
+	event->position = selector->position;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value = 0;
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int err = 0;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+				    sel_pins[i].pin,
+				    &value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			break;
+		}
+
+		if (value) {
+			selector->position = i;
+			break;
+		}
+	}
+
+	if (err) {
+		return err;
+	}
+
+	if (!value) {
+		selector->position = DEFAULT_POSITION;
+		LOG_WRN("Selector %" PRIu8 " state is unknown",
+			selector->config->id);
+	}
+
+	selector_event_send(selector);
+
+	return 0;
+}
+
+static int enable_interrupts_lock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int key = irq_lock();
+	int err = 0;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		if (i == selector->position) {
+			continue;
+		}
+
+		err = gpio_pin_enable_callback(gpio_dev[sel_pins[i].port],
+					       sel_pins[i].pin);
+		if (err) {
+			LOG_ERR("Cannot enable callback (err %d)", err);
+			break;
+		}
+	}
+
+	irq_unlock(key);
+
+	return err;
+}
+
+static int disable_interrupts_nolock(struct selector *selector)
+{
+	const struct gpio_pin *sel_pins = selector->config->pins;
+	int err = 0;
+
+	for (size_t i = 0; i < selector->config->pins_size; i++) {
+		err = gpio_pin_disable_callback(gpio_dev[sel_pins[i].port],
+						sel_pins[i].pin);
+
+		if (err) {
+			LOG_ERR("Cannot disable callback (err %d)", err);
+			break;
+		}
+	}
+
+	return err;
+}
+
+static void selector_isr(struct device *dev, struct gpio_callback *cb,
+			 u32_t pins_mask)
+{
+	u8_t port = dev - gpio_dev[0];
+	struct selector *sel;
+
+	sel = CONTAINER_OF((u8_t *)cb - port * sizeof(struct gpio_callback),
+			   struct selector,
+			   gpio_cb);
+	disable_interrupts_nolock(sel);
+
+	k_delayed_work_submit(&sel->work, DEBOUNCE_DELAY);
+}
+
+static int read_state_and_enable_interrupts(struct selector *selector)
```

Comment:
Well no, change in name is not solution here because it states even stronger that the function violates single reposnisbility principle. I suggest to direct use of functions called inside the function.

Reply:
It was added to remove code duplication